sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala

```diff
@@ -316,6 +362,86 @@ object RewriteDistinctAggregates extends Rule[LogicalPlan] {
+            // If(id > 1) 'a else null first, and use the result as output.
+            // Second, If more than one DISTINCT aggregate expression uses the same column,
+            // We need to construct the phantom attributes so as the output not lost.
+            // e.g. SUM (DISTINCT a), COUNT (DISTINCT a) FILTER (WHERE id > 1) will output
+            // attribute '_gen_distinct-1 and attribute '_gen_distinct-2 instead of two 'a.
+            // Note: We just need to illusion the expression with filter clause.
+            // The illusionary mechanism may result in multiple distinct aggregations uses
+            // different column, so we still need to call `rewrite`.
+            val phantomId = NamedExpression.newExprId.id
+            val unfoldableChildren = af.children.filter(!_.foldable)
+            val exprAttrs = unfoldableChildren.map { e =>
+              (e, AttributeReference(s"_gen_distinct_$phantomId", e.dataType, nullable = true)())
+            }
+            val exprAttrLookup = exprAttrs.toMap
+            val newChildren = af.children.map(c => exprAttrLookup.getOrElse(c, c))
+            val raf = af.withNewChildren(newChildren).asInstanceOf[AggregateFunction]
+            val aggExpr = ae.copy(aggregateFunction = raf, filter = None)
+            // Expand projection
+            val projection = unfoldableChildren.map {
+              case e if filter.isDefined => If(filter.get, e, nullify(e))
+              case e => e
+            }
+            (projection, exprAttrs, (ae, aggExpr))
+        }
+        val rewriteDistinctAttrMap = rewriteDistinctOperatorMap.flatMap(_._2)
+        val distinctAggChildAttrs = rewriteDistinctAttrMap.map(_._2)
+        val allAggAttrs = regularAggChildAttrMap.map(_._2) ++ distinctAggChildAttrs
+        // Construct the aggregate input projection.
+        val rewriteDistinctProjections = rewriteDistinctOperatorMap.flatMap(_._1)
+        val rewriteAggProjections =
+          Seq((a.groupingExpressions ++ regularAggChildren ++ rewriteDistinctProjections))
+        val groupByMap = a.groupingExpressions.collect {
+          case ne: NamedExpression => ne -> ne.toAttribute
+          case e => e -> AttributeReference(e.sql, e.dataType, e.nullable)()
+        }
+        val groupByAttrs = groupByMap.map(_._2)
+        // Construct the expand operator.
+        val expand = Expand(rewriteAggProjections, groupByAttrs ++ allAggAttrs, a.child)
+        val rewriteAggExprLookup =
+          (rewriteDistinctOperatorMap.map(_._3) ++ regularOperatorMap).toMap
+        val patchedAggExpressions = a.aggregateExpressions.map { e =>
+          e.transformDown {
+            case ae: AggregateExpression => rewriteAggExprLookup.getOrElse(ae, ae)
+          }.asInstanceOf[NamedExpression]
+        }
+        val expandAggregate = Aggregate(groupByAttrs, patchedAggExpressions, expand)
+        if (distinctAggExpressions.size > 1) {
+          rewrite(expandAggregate)
```

Comment:
Can you avoid this recursive call? We cannot move this pre-processing logic (pushing down aggregate filters into projections in `Expand`) into the line 186? https://github.com/apache/spark/blob/1846b0261b84ce1bca079bc59fb4518bff910c18/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDistinctAggregates.scala#L186

Reply:
It seems better. I will try.

Reply:
Thanks!